src/crimson/net/ProtocolV2.cc

```diff
@@ -0,0 +1,1603 @@
+                                          connect_seq,
+                                          conn.in_seq);
+  logger().debug("{} reconnect to session: client_cookie={}"
+                 " server_cookie={} gs={} cs={} ms={}",
+                 conn, client_cookie, server_cookie,
+                 global_seq, connect_seq, conn.in_seq);
+  return write_frame(reconnect)
+  .then([this] {
+    return read_main_preamble();
+  }).then([this] (Tag tag) {
+    switch (tag) {
+      case Tag::SESSION_RETRY_GLOBAL:
+        return read_frame_payload()
+        .then([this] {
+          // handle_session_retry_global() logic
+          auto retry = RetryGlobalFrame::Decode(rx_segments_data.back());
+          global_seq = messenger.get_global_seq(retry.global_seq());
+          logger().warn("{} received session retry global "
+                        "global_seq={}, choose new gs={}",
+                        conn, retry.global_seq(), global_seq);
+          return client_reconnect();
+        });
+      case Tag::SESSION_RETRY:
+        return read_frame_payload()
+        .then([this] {
+          // handle_session_retry() logic
+          auto retry = RetryFrame::Decode(rx_segments_data.back());
+          connect_seq = retry.connect_seq() + 1;
+          logger().warn("{} received session retry connect_seq={}, inc to cs={}",
+                        conn, retry.connect_seq(), connect_seq);
+          return client_reconnect();
+        });
+      case Tag::SESSION_RESET:
+        return read_frame_payload()
+        .then([this] {
+          // handle_session_reset() logic
+          auto reset = ResetFrame::Decode(rx_segments_data.back());
+          logger().warn("{} received session reset full={}", reset.full());
+          reset_session(reset.full());
+          return client_connect();
+        });
+      case Tag::WAIT:
+        return process_wait();
+      case Tag::SESSION_RECONNECT_OK:
+        return read_frame_payload()
+        .then([this] {
+          // handle_reconnect_ok() logic
+          auto reconnect_ok = ReconnectOkFrame::Decode(rx_segments_data.back());
+          logger().debug("{} received reconnect ok:"
+                         "sms={}, lossy={}, features={}",
+                         conn, reconnect_ok.msg_seq(),
+                         connect_seq, conn.policy.lossy, conn.features);
+          // TODO
+          // discard_requeued_up_to()
+          // backoff = utime_t();
+          return dispatcher.ms_handle_connect(
+              seastar::static_pointer_cast<SocketConnection>(
+                conn.shared_from_this()))
+          .handle_exception([this] (std::exception_ptr eptr) {
+            logger().error("{} ms_handle_connect caust exception: {}", conn, eptr);
+          });
+        }).then([this] {
+          return true;
+        });
+      default: {
+        unexpected_tag(tag, conn, "post_client_reconnect");
+        // won't be executed
+        return seastar::make_ready_future<bool>(false);
```

Comment:
nit, i am not sure the compiler will complain if we don't return any value here. but if we want to be on the safe side. probably, we could just "return false"? to be consistent with what we have 4 lines above?

Reply:
compiler will also complain if we just `return false` there.